### PR TITLE
website/blogs: Blog dockers

### PR DESCRIPTION
**Pull request by tanberry**
Blog about Dockers

### CODE REVIEW

File: website/blog/2023-10-26-you-might-be-doing-containers-wrong/item.md [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+---
+
+_authentik is an open source Identity Provider that unifies your identity needs into a single platform, replacing Okta, Active Directory, and Auth0. Authentik Security is a [public benefit company](https://github.com/OpenCoreVentures/ocv-public-benefit-company/blob/main/ocv-public-benefit-company-charter.md) building on top of the open source project._
+
+---
+
+There are two ways to judge an application:
+
+1. Does it do what it’s supposed to do?
+2. How easy is it to run?
```
Comment (kensternberg-authentik) [inline website/blog/2023-10-26-you-might-be-doing-containers-wrong/item.md:29]: Suggestion: "Is it easy to run?" preserves parallelism and improves impact.

File: website/blog/2023-10-26-you-might-be-doing-containers-wrong/item.md [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+Of course, your application architecture and deployment model need to support this setup, which is why it’s critical to think about these things when you’re starting out. If you’re reading this and thinking, I have a small-scale, hobby project, this doesn’t apply to me—let me put it this way: you will never regret setting things up the “right” way. It’s not going to come back to bite you if your situation changes later. It also gives users who install the application a lot more freedom and flexibility in how _they_ want to run it.
+
+## Mistake #2: Installing things at runtime
+
+Your container image should be complete in itself: it should contain all code and dependencies—everything it needs to run. This is the point of a container—it’s self contained.
+
+I’ve seen people set up their container to download an application from the vendor and install it into the container on startup. While this does work, what happens if you don’t have internet access? What if the vendor shuts down?
```
Comment (kensternberg-authentik) [inline website/blog/2023-10-26-you-might-be-doing-containers-wrong/item.md:73]: "What if the vendor shut down and that URL now points to a malicious bit of code?"  :-)

File: website/blog/2023-10-26-you-might-be-doing-containers-wrong/item.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+### Instead: Pin your dependencies
+
+You should be pinning your dependencies to a specific version, and updating to newer versions intentionally rather than by default.
+
+The most reliable way to do this is with a process called GitOps:
+
+-   In the context of Kubernetes, all the YAML files you deploy with Kubernetes are stored in the central Git repository.
+-   You have software in your Kubernetes cluster that automatically pulls the files from your Git repo and installs them into the cluster.
+-   Then you can use a tool like [Dependabot](https://github.com/dependabothttps://github.com/dependabot) or [Renovate](https://github.com/renovatebot/renovate) to automatically create PRs with a new version (if there is one) so you can test and approve it, and it’s all captured in your Git history.
```
Comment (kensternberg-authentik) [inline website/blog/2023-10-26-you-might-be-doing-containers-wrong/item.md:93]: Typo: That first URL looks munged.

File: website/blog/2023-10-26-you-might-be-doing-containers-wrong/item.md [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+GitOps might be a bit excessive if you’re just running one small hobby project on a single server, but in any case you should still pin a version.
```
Comment (kensternberg-authentik) [inline website/blog/2023-10-26-you-might-be-doing-containers-wrong/item.md:95]: Suggestion: "... only running a small hobby project ..."

File: website/blog/2023-10-26-you-might-be-doing-containers-wrong/item.md [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+We have since added it due to popular request. This is how authentik’s version tags work:
+
+-   Our version number is 3 digits reflecting the date of the release, so the latest currently is [2023.10.1](https://goauthentik.io/docs/releases/2023.10).
+    -   You can either use 2023.10.1 as the tag, pinning to that specific version
+    -   You can pin to 2023.10, which you means that you always get the latest patch version, or
+    -   You can use 2023, which means you always get the latest version within that year.
+
+The principle is roughly the same with any project using [SemVer](https://semver.org/): you could just lock to v1, which means you get the latest v1 with all minor patches and fixes, without breaking updates. Then you switch to v2 when you’re ready.
+
+With this approach you are putting some trust in the developer not to publish any breaking changes with the wrong version number (but you’re technically always putting trust in some developer when using someone else’s software!).
```
Comment (kensternberg-authentik) [inline website/blog/2023-10-26-you-might-be-doing-containers-wrong/item.md:108]: Suggestion: "...not to publish breaking changes..."  The any is superfluous.

File: website/blog/2023-10-26-you-might-be-doing-containers-wrong/item.md [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+## Mistake #3: Writing logs to files instead of stdout
+
+This is another issue on the infrastructure side that mainly happens when you put legacy applications into containers. It used to be standard that applications put their log output into a file, and you’d probably have a system daemon set up to rotate those files and archive the old ones. This was great when everything ran on the same server without containers.
+
+A lot of software still does that, but sending log outputs to files makes it very hard to integrate with anything because, with Docker (and containers in general), there is the expectation that you log to the standard output (`stdout`):
```
Comment (kensternberg-authentik) [inline website/blog/2023-10-26-you-might-be-doing-containers-wrong/item.md:114]: Comment: As an old skool developer, the "integrate with anything" is confusing.  

Suggestion: "A lot of software still logs to files by default, but this makes collecting and aggregating your services logs much harder. Docker (and containers in general) expect that you log to standard output so your orchestration platform can route the logs to your monitoring tool of choice."

File: website/blog/2023-10-26-you-might-be-doing-containers-wrong/item.md [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+Docker puts the logs into a JSON file that it can read itself and see the timestamps and which container the log refers to. You can set up log forwarding with both Docker and Kubernetes. If you have a central logging server, the plugin gets the standard output of a container and sends it to that server.
```
Comment (kensternberg-authentik) [inline website/blog/2023-10-26-you-might-be-doing-containers-wrong/item.md:116]: Suggestion: Links to documentation for "log forwarding with Docker" and "log forwarding with Kubernetes."

**Review by kensternberg-authentik — APPROVED** [review]
Before committing: there is one typo (a bad URL; looks like someone hit `paste` twice) that I believe must be fixed.

I have otherwise appended a few suggestions and comments, but otherwise this looks good.  It has a strong Jens voice to it.  :-)

File: website/blog/2023-10-26-you-might-be-doing-containers-wrong/item.md (hunk 1 above)
Comment (tanberry) [reply website/blog/2023-10-26-you-might-be-doing-containers-wrong/item.md:29]: Parallelism should never be underestimated.

File: website/blog/2023-10-26-you-might-be-doing-containers-wrong/item.md (hunk 7 above)
Comment (tanberry) [reply website/blog/2023-10-26-you-might-be-doing-containers-wrong/item.md:116]: Do we have docs on that? I searched.. will add links later if I can find them!

**Pull request merged**